File: src/text.ts
import type { JsonTileDelimiter } from './types.js';

export function escapeTileText(
  value: string,
  delimiter: string = '\t'
): string {
  let result = '';

  for (const char of value) {
    if (char === '\\') {
      result += '\\\\';
    } else if (char === '\t') {
      result += '\\t';
    } else if (char === '\n') {
      result += '\\n';
    } else if (char === '\r') {
      result += '\\r';
    } else if (char === ' ' && delimiter === ' ') {
      result += '\\s';
    } else if (char === delimiter && delimiter !== '\t') {
      result += `\\${char}`;
    } else {
      result += char;
    }
  }

  return result;
}

export function unescapeTileText(value: string): string {
  let result = '';

  for (let index = 0; index < value.length; index += 1) {
    const char = value[index] ?? '';
    if (char !== '\\') {
      result += char;
      continue;
    }

    const escaped_char = value[index + 1];
    if (escaped_char === undefined) {
      throw new Error('Invalid TILE escape: trailing backslash');
    }

    if (escaped_char === 't') {
      result += '\t';
    } else if (escaped_char === 'n') {
      result += '\n';
    } else if (escaped_char === 'r') {
      result += '\r';
    } else if (escaped_char === '\\') {
      result += '\\';
    } else if (escaped_char === 's') {
      result += ' ';
    } else if (escaped_char === ',') {
      result += ',';
    } else if (escaped_char === '|') {
      result += '|';
    } else if (escaped_char === '@') {
      result += '@';
    } else if (escaped_char === ':') {
      result += ':';
    } else {
      throw new Error(`Invalid TILE escape: \\${escaped_char}`);
    }

    index += 1;
  }

  return result;
}

export function resolveTileDelimiter(delimiter: JsonTileDelimiter): string {
  if (delimiter === 'comma') {
    return ',';
  }

  if (delimiter === 'pipe') {
    return '|';
  }

  if (delimiter === 'space') {
    return ' ';
  }

  return '\t';
}

File: packages/python/src/software_dlc_tile/text.py
from __future__ import annotations


def escape_tile_text(value: str, delimiter: str = "\t") -> str:
    result: list[str] = []

    for char in value:
        if char == "\\":
            result.append("\\\\")
        elif char == "\t":
            result.append("\\t")
        elif char == "\n":
            result.append("\\n")
        elif char == "\r":
            result.append("\\r")
        elif char == " " and delimiter == " ":
            result.append("\\s")
        elif char == delimiter and delimiter != "\t":
            result.append(f"\\{char}")
        else:
            result.append(char)

    return "".join(result)


def unescape_tile_text(value: str) -> str:
    result: list[str] = []
    index = 0

    while index < len(value):
        char = value[index]
        if char != "\\":
            result.append(char)
            index += 1
            continue

        if index + 1 >= len(value):
            raise ValueError("Invalid TILE escape: trailing backslash")

        escaped_char = value[index + 1]
        if escaped_char == "t":
            result.append("\t")
        elif escaped_char == "n":
            result.append("\n")
        elif escaped_char == "r":
            result.append("\r")
        elif escaped_char == "\\":
            result.append("\\")
        elif escaped_char == "s":
            result.append(" ")
        elif escaped_char == ",":
            result.append(",")
        elif escaped_char == "|":
            result.append("|")
        elif escaped_char == "@":
            result.append("@")
        elif escaped_char == ":":
            result.append(":")
        else:
            raise ValueError(f"Invalid TILE escape: \\{escaped_char}")

        index += 2

    return "".join(result)


def resolve_tile_delimiter(delimiter: str) -> str:
    if delimiter == "comma":
        return ","

    if delimiter == "pipe":
        return "|"

    if delimiter == "space":
        return " "

    return "\t"
